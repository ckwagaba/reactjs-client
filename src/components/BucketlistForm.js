import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Nav from './Nav';
import SubmitButton from './SubmitButton';
import ServerResponse from './ServerResponse';
import { BASEURL } from '../Config.js';
import store from '../store/Store';

class BucketlistForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      APIResponse: '',
      name: '',
      buttonText: 'Add'
    }
  }

componentDidMount() {
  // load with a name value for update case
  if(this.props.match.params.bucketlistName){
    this.setState(
      {
        name: this.props.match.params.bucketlistName,
        buttonText: 'Update'
      }
    );
  }
}

  // handleNameInput
  handleNameInput = (event) => {
    this.setState(
      {
        name: event.target.value
      }
    );
  }

  // create or update bucketlist
  createOrUpdateBucketlist = (requestData, url, method) => {
    fetch(url, {
      method: method,
      body: JSON.stringify(requestData),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': store.getState().authView.ACCESSTOKEN
      }
    })
    .then(response => {
      return response.json()
    })
    .then(responseData => {
      // display message in case of an error
      if(responseData.message) {
        this.setState(
          {
            APIResponse: responseData.message,
            name: ''
          }
        );
      }
      else {
        this.props.history.push('/bucketlists/');
      }
    });
  }

  // add bucketlist
  addBucketlist = (event) => {
    event.preventDefault();
    const requestData = {
      'name': this.state.name
    };
    const ENDPOINT = BASEURL + '/bucketlists/';
    // for an update
    if(this.props.match.params.bucketlistId){
      this.createOrUpdateBucketlist(requestData, ENDPOINT + this.props.match.params.bucketlistId, 'PUT');
    }
    // for a creation
    else {
      this.createOrUpdateBucketlist(requestData, ENDPOINT, 'POST');
    }
  }

  render () {
    return (
      <div className="landing_page bucketlist_view">
        <Nav handleLogout={this.props.handleLogout} />
        <header>
          <div className="header_left">
            <span>Bucketlist</span>
          </div>
          <div className="header_right">
            <Link to="/bucketlists/">
              <div className="round_button cancel_button">+</div>
            </Link>
          </div>
        </header>
        <main className="entry_form">
          <form onSubmit={this.addBucketlist}>
            <ServerResponse serverResponse={this.state.APIResponse} />
            <input type="text" className="text_input" placeholder="Bucketlist Name" value={this.state.name} onChange={this.handleNameInput} />
            <SubmitButton buttonText={this.state.buttonText} />
          </form>
        </main>
      </div>
    );
  }
}

export default BucketlistForm;
