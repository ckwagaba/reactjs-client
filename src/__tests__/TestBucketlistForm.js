import React from 'react';
import { shallow } from 'enzyme';
import BucketlistForm from '../components/BucketlistForm';

describe('BucketlistForm component', () => {

  const props = {
    match: {
      params: {
        bucketlistId: 1,
        itemName: 'career',
        itemId: 1
      }
    },
    history: '/bucketlists/'
  }

  const component = shallow(
    <BucketlistForm {...props} />
  );

  it('should exist', () => {
    expect(component).toMatchSnapshot();
  });

});
