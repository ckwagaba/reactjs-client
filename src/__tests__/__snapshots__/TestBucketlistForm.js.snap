// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BucketlistForm component should exist 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BucketlistForm
    history="/bucketlists/"
    match={
        Object {
            "params": Object {
              "bucketlistId": 1,
              "itemId": 1,
              "itemName": "career",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withRouter(Nav)
          handleLogout={undefined}
/>,
        <header>
          <div
                    className="header_left"
          >
                    <span>
                              Bucketlist
                    </span>
          </div>
          <div
                    className="header_right"
          >
                    <Link
                              replace={false}
                              to="/bucketlists/"
                    >
                              <div
                                        className="round_button cancel_button"
                              >
                                        +
                              </div>
                    </Link>
          </div>
</header>,
        <main
          className="entry_form"
>
          <form
                    onSubmit={[Function]}
          >
                    <ServerResponse
                              serverResponse=""
                    />
                    <input
                              className="text_input"
                              onChange={[Function]}
                              placeholder="Bucketlist Name"
                              type="text"
                              value=""
                    />
                    <SubmitButton
                              buttonText="Add"
                    />
          </form>
</main>,
      ],
      "className": "landing_page bucketlist_view",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleLogout": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="header_left"
>
              <span>
                            Bucketlist
              </span>
</div>,
            <div
              className="header_right"
>
              <Link
                            replace={false}
                            to="/bucketlists/"
              >
                            <div
                                          className="round_button cancel_button"
                            >
                                          +
                            </div>
              </Link>
</div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <span>
                Bucketlist
</span>,
              "className": "header_left",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Bucketlist",
              },
              "ref": null,
              "rendered": "Bucketlist",
              "type": "span",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Link
                replace={false}
                to="/bucketlists/"
>
                <div
                                className="round_button cancel_button"
                >
                                +
                </div>
</Link>,
              "className": "header_right",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <div
                  className="round_button cancel_button"
>
                  +
</div>,
                "replace": false,
                "to": "/bucketlists/",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "+",
                  "className": "round_button cancel_button",
                },
                "ref": null,
                "rendered": "+",
                "type": "div",
              },
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <form
            onSubmit={[Function]}
>
            <ServerResponse
                        serverResponse=""
            />
            <input
                        className="text_input"
                        onChange={[Function]}
                        placeholder="Bucketlist Name"
                        type="text"
                        value=""
            />
            <SubmitButton
                        buttonText="Add"
            />
</form>,
          "className": "entry_form",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ServerResponse
                serverResponse=""
/>,
              <input
                className="text_input"
                onChange={[Function]}
                placeholder="Bucketlist Name"
                type="text"
                value=""
/>,
              <SubmitButton
                buttonText="Add"
/>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "serverResponse": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "text_input",
                "onChange": [Function],
                "placeholder": "Bucketlist Name",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "buttonText": "Add",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "form",
        },
        "type": "main",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withRouter(Nav)
            handleLogout={undefined}
/>,
          <header>
            <div
                        className="header_left"
            >
                        <span>
                                    Bucketlist
                        </span>
            </div>
            <div
                        className="header_right"
            >
                        <Link
                                    replace={false}
                                    to="/bucketlists/"
                        >
                                    <div
                                                className="round_button cancel_button"
                                    >
                                                +
                                    </div>
                        </Link>
            </div>
</header>,
          <main
            className="entry_form"
>
            <form
                        onSubmit={[Function]}
            >
                        <ServerResponse
                                    serverResponse=""
                        />
                        <input
                                    className="text_input"
                                    onChange={[Function]}
                                    placeholder="Bucketlist Name"
                                    type="text"
                                    value=""
                        />
                        <SubmitButton
                                    buttonText="Add"
                        />
            </form>
</main>,
        ],
        "className": "landing_page bucketlist_view",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleLogout": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="header_left"
>
                <span>
                                Bucketlist
                </span>
</div>,
              <div
                className="header_right"
>
                <Link
                                replace={false}
                                to="/bucketlists/"
                >
                                <div
                                                className="round_button cancel_button"
                                >
                                                +
                                </div>
                </Link>
</div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <span>
                  Bucketlist
</span>,
                "className": "header_left",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Bucketlist",
                },
                "ref": null,
                "rendered": "Bucketlist",
                "type": "span",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/bucketlists/"
>
                  <div
                                    className="round_button cancel_button"
                  >
                                    +
                  </div>
</Link>,
                "className": "header_right",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <div
                    className="round_button cancel_button"
>
                    +
</div>,
                  "replace": false,
                  "to": "/bucketlists/",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "+",
                    "className": "round_button cancel_button",
                  },
                  "ref": null,
                  "rendered": "+",
                  "type": "div",
                },
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <form
              onSubmit={[Function]}
>
              <ServerResponse
                            serverResponse=""
              />
              <input
                            className="text_input"
                            onChange={[Function]}
                            placeholder="Bucketlist Name"
                            type="text"
                            value=""
              />
              <SubmitButton
                            buttonText="Add"
              />
</form>,
            "className": "entry_form",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <ServerResponse
                  serverResponse=""
/>,
                <input
                  className="text_input"
                  onChange={[Function]}
                  placeholder="Bucketlist Name"
                  type="text"
                  value=""
/>,
                <SubmitButton
                  buttonText="Add"
/>,
              ],
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "serverResponse": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "text_input",
                  "onChange": [Function],
                  "placeholder": "Bucketlist Name",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "buttonText": "Add",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "form",
          },
          "type": "main",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
